### PR TITLE
add support for jsonUnknown trait

### PR DESCRIPTION
**Pull request by benoitlouy**
Implement support for `jsonUnknown` trait defined in https://github.com/disneystreaming/alloy/pull/180

### CODE REVIEW

File: modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala [hunk 1]
```diff
@@ -1445,14 +1480,21 @@ private[smithy4s] class SchemaVisitorJCodec(
           cursor: Cursor,
           in: JsonReader
       ): scala.collection.Map[String, Any] => Z = {
+        val unknownValues =
+          if (keepUnknown) ListBuffer[(String, Document)]() else null
```
Comment (Baccata) [inline modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala]: I appreciate the effort to minimise allocations :+1: . I would personally have gone for distinct `JCodec` implementations based on whether the structure has a `@jsonUnknown` field or not, which does induce some copy/pasting, but completely shaves off some checks on the critical path.

File: modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala [hunk 2]
```diff
@@ -1467,14 +1509,32 @@ private[smithy4s] class SchemaVisitorJCodec(
...
-                else default
-              } else value
+              if (isLabelForUnknownFields(jsonLabel)) {
+                if (unknownValue == null) {
+                  if (default == null) {
+                    if (f.isRequired) Map.empty
+                    else None
+                  } else default
+                } else {
+                  if (f.isRequired) unknownValue else Some(unknownValue)
```
Comment (Baccata) [inline modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala]: you cannot make the assumption that `!required` implies `scala.Option` : we're planning on relaxing our Schema model to allow for other ways of modelling optionality (to cater for `java.util.Optional`, for instance). 

Moreover, maybe the user applies other type of customisations to what ends up annotated with `jsonUnknown`. Maybe they have a bijection from Document to circe's Json in place, etc. 

So what you need to do instead is to compile the `Document.Decoder` associated to the field that is annotated with `@jsonUnkown`, in order to run the unknownValue (converted to `Document`) through it to get a value that'll be guaranteed to be of the correct type, without having to make any assumption.

File: modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala [hunk 3]
```diff
@@ -1467,14 +1509,32 @@ private[smithy4s] class SchemaVisitorJCodec(
...
-              if (value == null) {
-                if (default == null)
-                  cursor.requiredFieldError(jsonLabel, jsonLabel)
-                else default
-              } else value
+              if (isLabelForUnknownFields(jsonLabel)) {
+                if (unknownValue == null) {
+                  if (default == null) {
+                    if (f.isRequired) Map.empty
+                    else None
```
Comment (Baccata) [inline modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala]: See my comment [here](https://github.com/disneystreaming/smithy4s/pull/1574/files#r1713392500)

File: modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala (hunk 1 above)
Comment (benoitlouy) [reply modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala]: There are now 2 `JCodec` implementations to separately handle structs with a `jsonUnknown` member and ones without.

File: modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala (hunk 2 above)
Comment (benoitlouy) [reply modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala]: This has been addressed and the decoding now uses what you described.

File: modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala [hunk 4]
```diff
@@ -1415,7 +1436,122 @@ private[smithy4s] class SchemaVisitorJCodec(
...
+                  } else value
+
+                case Some(docDecoder) =>
+                  if (unknownValue == null) {
+                    if (default == null) {
+                      docDecoder
+                        .decode(Document.obj())
+                        .getOrElse(
+                          throw new RuntimeException(
+                            s"${cursor.getPath(Nil)} Failed translating a Document.DObject to the type targeted by ${f.label}."
```
Comment (Baccata) [inline modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala:1529]: use `in.decodeError` instead, to throw the exception

File: modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala [hunk 5]
```diff
@@ -1415,7 +1436,122 @@ private[smithy4s] class SchemaVisitorJCodec(
...
+                          )
+                        )
+                    } else default
+                  } else {
+                    docDecoder
+                      .decode(unknownValue)
+                      .getOrElse(
+                        throw new RuntimeException(
+                          s"${cursor.getPath(Nil)} Failed translating a Document.DObject to the type targeted by ${f.label}."
```
Comment (Baccata) [inline modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala:1538]: [ditto](https://github.com/disneystreaming/smithy4s/pull/1574/files#r1714988906)

File: modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala [hunk 6]
```diff
@@ -1384,27 +1388,44 @@ private[smithy4s] class SchemaVisitorJCodec(
...
+      case m: Map[_, _] =>
+        m.foreach { case (label: String, value: Document) =>
+          writeLabel(label, out)
+          documentJCodec.encodeValue(value, out)
+        }
+      case Some(m: Map[_, _]) =>
+        m.foreach { case (label: String, value: Document) =>
+          writeLabel(label, out)
+          documentJCodec.encodeValue(value, out)
+        }
```
Comment (Baccata) [inline modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala]: Sorry, I should have signposted this bit as well, but this is falling under the similar notion to what I was describing [here](https://github.com/disneystreaming/smithy4s/pull/1574#discussion_r1713392500) : 

1. You cannot make the assumption that a smithy map necessarily translates to a Scala map 
2. You cannot make the assumption that the absence of `required` translates to Scala Option 
3. You cannot even make the assumption that at this level, you're dealing with `Documents`

Therefore you need to pre-compile the `Document.Encoder` and encode the field value against it, and pattern-match against the produced document to check it's a DObject, in which case you run the exact same `foreach` you have there

**Comment by Baccata** [comment]
This is shaping up well ! This code is quite tricky and it's a very valuable PR, thank you very much for taking this on 👍

File: modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala (hunk 6 above)
Comment (benoitlouy) [reply modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala]: This makes sense, thanks for the directions.

**Review by Baccata — APPROVED** [review]
Giving a pre-emptive approval for this, for @lewisjkl to consider when the upstream work (alloy) has been merged and released

**Review by lewisjkl — APPROVED** [review]
This looks good to me as well, great work on this one!

**Review by Baccata — APPROVED** [review]

**Pull request merged**